# Examen 2: Métodos computacionales
A)
1. Usando su número de cédula cómo semilla, obtenga un numero entero aleatorio, $n$, entre 2 y 20.
2. Usando el $n$ obtenido, haga una estimación de la siguiente integral hasta dos cifras significativas, usando el método `quad` del módulo `integrate` de Scipy
$$
\int_0^\infty \sin(x^n)\operatorname{d}x
$$
Sugerencia. Compare con el valor obtenido desde https://www.wolframalpha.com/ o [sympy](https://stackoverflow.com/questions/65862611/integration-of-sin-x-dx-problem-in-python-programming-implementation)

In [82]:
import numpy as np
from scipy import interpolate
import scipy.integrate as integrate
from math import sin
from math import log

In [83]:
np.random.seed(1001477629)
n=np.random.uniform(2,20)
n

8.28158661644435

In [88]:
f1=lambda x: sin(1/x**n)/x**2


In [101]:
I1=integrate.quad(f1,0.5,np.inf)
I1[0]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: IntegrationWarning: The maximum number of subdivisions (50) has been achieved.
  If increasing the limit yields no improvement it is advised to analyze 
  the integrand in order to determine the difficulties.  If the position of a 
  local difficulty can be determined (singularity, discontinuity) one will 
  probably gain from splitting up the interval and calling the integrator 
  on the subranges.  Perhaps a special-purpose integrator should be used.
  """Entry point for launching an IPython kernel.


0.17861676778256674

$$
I_1=0.17861676778256674
$$
$$
\int_0^{0.5} f_1(x)\operatorname{d}y   = -0.000766728
$$
$$
\int_0^\infty \sin(x^n)\operatorname{d}x ≈ I_1 - 0.000766728 = 0.17785 ≈ 0.17862
$$



B) Cargue los datos disponibles en:

[https://github.com/ComputationalMethods/Evaluacion_2022-1/blob/main/data/lista_pares.json](https://raw.githubusercontent.com/ComputationalMethods/Evaluacion_2022-1/main/data/lista_pares.json)

como un DataFrame de pandas. Este contendrá dos columnas. La columna `lista`, con una lista de 8 enteros (algunos de ellos repetidos) y una columna `pares` correspondiente a un diccionario
con dos items: un entero rotulado con `'S'` y una lista de pares de enteros routulada con `'ψ'`. Note que el resultado de la suma en valor absoluto de los dos enteros en cada uno de los pares reproduce el valor del correspondiente `'S'` 

Seleccione la fila del DataFrame con el índice correspondiente a los dos últimos dígitos de su cédula y desarrolle los siguientes puntos con la correspondiente  `lista` y `pares`:
1. Construya una matrix $8\times 8$ _simétrica_ donde cada fila y la correspondiente columna correspondan a un entero de la `lista` (incluyendo los repetidos); de manera que en los cruces de filas y columnas definidos por los pares de enteros de `'ψ'`, se genere una entrada aleatoria mayor que cero y menor o igual que 1, con las demas entradas iguales a cero.
1. Calcule el determinante y establezca explícitamente si es diferente cero (tenga en cuenta la precisión numérica). Recuerde que si el determinante es cero, se espera que al menos uno de los autovalores sea cero
2. Organize los autovalores y autovectores de menor a mayor en valor absoluto y compruebe que las matriz de autovectores organizada da lugar a los autovalores en el orden correcto. 
3. Establezca explícitamente _cuantos_ autovalores son degenerados en valor absoluto.

In [11]:
import pandas as pd
import numpy as np
import sympy as sy
import scipy.linalg as linalg
import scipy

In [13]:
DF=pd.read_json('lista_pares.json',orient='records',lines=True)
Daniel=DF[29][0]
Daniel

{'lista': [1, -3, -6, -6, 7, 8, 11, -12],
 'pares': {'S': 5,
  'ψ': [[1, -6],
   [-3, 8],
   [7, -12],
   [-6, 11],
   [-6, 1],
   [8, -3],
   [-12, 7],
   [11, -6]]}}

In [14]:
M=([[0,0,1,1,0,0,0,0],
    [0,0,0,0,0,1,0,0],
    [1,0,0,0,0,0,1,0],
    [1,0,0,0,0,0,1,0],
    [0,0,0,0,0,0,0,1],
    [0,1,0,0,0,0,0,0],
    [0,0,1,1,0,0,0,0],
    [0,0,0,0,1,0,0,0]])

for i in range(0,8):
    for j in range(0,8):
        a=np.random.random()
        M[i][j]=M[i][j]*a
        M[j][i]=M[j][i]*a

sy.Matrix(M)

Matrix([
[                0.0,               0.0, 0.00915643985923252, 0.267530267288711,               0.0,               0.0,               0.0,               0.0],
[                0.0,               0.0,                 0.0,               0.0,               0.0, 0.204251304185014,               0.0,               0.0],
[0.00915643985923252,               0.0,                 0.0,               0.0,               0.0,               0.0, 0.528000893751763,               0.0],
[  0.267530267288711,               0.0,                 0.0,               0.0,               0.0,               0.0, 0.948603946126884,               0.0],
[                0.0,               0.0,                 0.0,               0.0,               0.0,               0.0,               0.0, 0.476507078921567],
[                0.0, 0.204251304185014,                 0.0,               0.0,               0.0,               0.0,               0.0,               0.0],
[                0.0,               0.0,   

In [15]:
Dm=np.linalg.det(M) #Determinante
Dm

0.00016647975455583885

In [16]:
l,V=np.linalg.eig(M)

In [17]:
V0=np.c_[ V[:,0] ]
V1=np.c_[ V[:,1] ]
V2=np.c_[ V[:,2] ]
V3=np.c_[ V[:,3] ]
V4=np.c_[ V[:,4] ]
V5=np.c_[ V[:,5] ]
V6=np.c_[ V[:,6] ]
V7=np.c_[ V[:,7] ]
V=np.c_[V0,V1,V2,V3,V4,V5,V6,V7]

In [18]:
sy.Matrix((V.transpose()@M@V).round(14))

Matrix([
[-1.11178789625787,              0.0,               0.0,              0.0,               0.0,              0.0,              0.0,               0.0],
[              0.0, 1.11178789625787,               0.0,              0.0,               0.0,              0.0,              0.0,               0.0],
[              0.0,              0.0, -0.11924071641485,              0.0,               0.0,              0.0,              0.0,               0.0],
[              0.0,              0.0,               0.0, 0.11924071641485,               0.0,              0.0,              0.0,               0.0],
[              0.0,              0.0,               0.0,              0.0, -0.20425130418501,              0.0,              0.0,               0.0],
[              0.0,              0.0,               0.0,              0.0,               0.0, 0.20425130418501,              0.0,               0.0],
[              0.0,              0.0,               0.0,              0.0,               0.

In [19]:
A_diag=np.diag(l)
sy.Matrix(A_diag)

Matrix([
[-1.11178789625787,              0.0,                0.0,               0.0,                0.0,               0.0,               0.0,                0.0],
[              0.0, 1.11178789625787,                0.0,               0.0,                0.0,               0.0,               0.0,                0.0],
[              0.0,              0.0, -0.119240716414849,               0.0,                0.0,               0.0,               0.0,                0.0],
[              0.0,              0.0,                0.0, 0.119240716414849,                0.0,               0.0,               0.0,                0.0],
[              0.0,              0.0,                0.0,               0.0, -0.204251304185014,               0.0,               0.0,                0.0],
[              0.0,              0.0,                0.0,               0.0,                0.0, 0.204251304185014,               0.0,                0.0],
[              0.0,              0.0,                0.

In [20]:
def argeig(A):
    λ,V=np.linalg.eig(A)
    V=np.c_[ tuple( [ np.c_[V[:,i]]    for i in np.abs(λ).argsort() ] ) ]
    λ=np.array([λ[i] for i in np.abs(λ).argsort()])
    return λ,V

In [21]:
ol,oV=argeig(M)

In [22]:
A_diag=np.diag(ol)
sy.Matrix(A_diag)

Matrix([
[-0.119240716414849,               0.0,                0.0,               0.0,                0.0,               0.0,               0.0,              0.0],
[               0.0, 0.119240716414849,                0.0,               0.0,                0.0,               0.0,               0.0,              0.0],
[               0.0,               0.0, -0.204251304185014,               0.0,                0.0,               0.0,               0.0,              0.0],
[               0.0,               0.0,                0.0, 0.204251304185014,                0.0,               0.0,               0.0,              0.0],
[               0.0,               0.0,                0.0,               0.0, -0.476507078921567,               0.0,               0.0,              0.0],
[               0.0,               0.0,                0.0,               0.0,                0.0, 0.476507078921567,               0.0,              0.0],
[               0.0,               0.0,                

In [23]:
sy.Matrix((oV.transpose()@M@oV).round(14))

Matrix([
[-0.11924071641485,              0.0,               0.0,              0.0,               0.0,              0.0,               0.0,              0.0],
[              0.0, 0.11924071641485,               0.0,              0.0,               0.0,              0.0,               0.0,              0.0],
[              0.0,              0.0, -0.20425130418501,              0.0,               0.0,              0.0,               0.0,              0.0],
[              0.0,              0.0,               0.0, 0.20425130418501,               0.0,              0.0,               0.0,              0.0],
[              0.0,              0.0,               0.0,              0.0, -0.47650707892157,              0.0,               0.0,              0.0],
[              0.0,              0.0,               0.0,              0.0,               0.0, 0.47650707892157,               0.0,              0.0],
[              0.0,              0.0,               0.0,              0.0,               0.